scipy/stats/_mannwhitneyu.py

```diff
@@ -35,143 +50,100 @@ def pmf(self, k, m, n):
-        if self._fmnks[m, n, k] >= 0:
-            return self._fmnks[m, n, k]
-
-        if k == 0 and m >= 0 and n >= 0:  # [3] Theorem 2.5 Line 2
-            fmnk = 1
-        else:   # [3] Theorem 2.5 Line 3 / Equation 3
-            fmnk = self._f(m-1, n, k-n) + self._f(m, n-1, k)
-
-        self._fmnks[m, n, k] = fmnk  # remember result
-
-        return fmnk
-
-
-# Maintain state for faster repeat calls to mannwhitneyu w/ method='exact'
-_mwu_state = _MWU()
-
-
-def _mwu_f_iterative(m, n, k, fmnks):
-    '''Iterative implementation of function of [3] Theorem 2.5'''
-
-    def _base_case(m, n, k):
-        '''Base cases from recursive version'''
-
-        # if already calculated, return the value
-        if fmnks.get((m, n, k), -1) >= 0:
-            return fmnks[(m, n, k)]
-
-        # [3] Theorem 2.5 Line 1
-        elif k < 0 or m < 0 or n < 0 or k > m*n:
-            return 0
-
-        # [3] Theorem 2.5 Line 2
-        elif k == 0 and m >= 0 and n >= 0:
-            return 1
-
-        return None
-
-    stack = [(m, n, k)]
-    fmnk = None
-
-    while stack:
-        # Popping only if necessary would save a tiny bit of time, but NWI.
-        m, n, k = stack.pop()
-
-        # If we're at a base case, continue (stack unwinds)
-        fmnk = _base_case(m, n, k)
-        if fmnk is not None:
-            fmnks[(m, n, k)] = fmnk
-            continue
-
-        # If both terms are base cases, continue (stack unwinds)
-        f1 = _base_case(m-1, n, k-n)
-        f2 = _base_case(m, n-1, k)
-        if f1 is not None and f2 is not None:
-            # [3] Theorem 2.5 Line 3 / Equation 3
-            fmnk = f1 + f2
-            fmnks[(m, n, k)] = fmnk
-            continue
-
-        # recurse deeper
-        stack.append((m, n, k))
-        if f1 is None:
-            stack.append((m-1, n, k-n))
-        if f2 is None:
-            stack.append((m, n-1, k))
-
-    return fmnks
+    # build_sigma_array and build_u_freqs_array adapted from code
+    # by @toobaz with permission. Thanks to @andreasloe for the suggestion.
+    # See https://github.com/scipy/scipy/pull/4933#issuecomment-1898082691
+    def build_sigma_array(self, a):
+        n1, n2 = self.n1, self.n2
+        if a + 1 <= self.s_array.size:
+            return self.s_array[1:a+1]
+
+        s_array = np.zeros(a + 1, dtype=int)
+
+        for d in np.arange(1, n1 + 1):
+            # All multiples of d, except 0:
+            indices = np.arange(d, a + 1, d)
+            # \epsilon_d = 1:
+            s_array[indices] += d
+
+        for d in np.arange(n2 + 1, n2 + n1 + 1):
+            # All multiples of d, except 0:
+            indices = np.arange(d, a + 1, d)
+            # \epsilon_d = -1:
+            s_array[indices] -= d
+
+        # We don't need 0:
+        self.s_array = s_array
+        return s_array[1:]
+
+    def build_u_freqs_array(self, maxu):
+        """
+        Build all the array of frequences for u from 0 to maxu.
```

Comment:
Typo?
```suggestion
        Build all the array of frequencies for u from 0 to maxu.
```

Reply:
Yup. I copied the code from https://github.com/scipy/scipy/pull/4933#issuecomment-1898082691 verbatim, making only necessary changes.